### PR TITLE
feat(connector): [Nexinets] add mandates and applepay support to nexinets

### PR DESCRIPTION
**Pull request by srujanchikke**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

1) Add support for card mandates for nexinets
2) Add apple pay wallet for nexinets

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

1)Apple pay testing is pending(Need to generate apple pay token)

2)Manual testing in postman for mandates
**Customer initiated transaction for 3ds card**
Screenshot 2023-04-27 at 12 07 21 PM
**payment retrieval for CIT**

Screenshot 2023-04-27 at 12 08 00 PM

**Successful merchant initiated transaction without challenge** 

Screenshot 2023-04-27 at 12 08 40 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/nexinets/transformers.rs [hunk 1]
```diff
@@ -42,25 +45,60 @@ pub enum NexinetsProduct {
...
+
+#[derive(Debug, Serialize)]
+#[serde(untagged)]
+pub enum CardDataDetails {
+    CardDetails(Box<CardDetails>),
+    PaymentInstrument(Box<PaymentInstrument>),
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
```
Comment (Abhicodes-crypto) [inline crates/router/src/connector/nexinets/transformers.rs:76]: The field are already in camel case. So this is not required I guess

File: crates/router/src/connector/nexinets/transformers.rs [hunk 2]
```diff
@@ -525,13 +586,88 @@ fn get_payment_details_and_product(
...
+        })))
+    }
+}
+
+fn get_payment_instrument(item: &types::PaymentsAuthorizeRouterData) -> PaymentInstrument {
+    let payment_instrument_id = item
+        .request
+        .mandate_id
+        .as_ref()
+        .and_then(|mandate_ids| mandate_ids.connector_mandate_id.clone());
```
Comment (jagan-jaya) [inline crates/router/src/connector/nexinets/transformers.rs:619]:
```suggestion
    let payment_instrument_id = item.request.connector_mandate_id();
```

File: crates/router/src/connector/nexinets/transformers.rs [hunk 3]
```diff
@@ -525,13 +586,88 @@ fn get_payment_details_and_product(
...
+        payment_instrument_id,
+    }
+}
+
+fn get_applepay_details(
+    wallet_data: &api_models::payments::WalletData,
+    applepay_data: &api_models::payments::ApplePayWalletData,
+) -> CustomResult<ApplePayDetails, errors::ConnectorError> {
+    let payment_data = wallet_data.get_wallet_token_as_json()?;
+    Ok(ApplePayDetails {
```
Comment (jagan-jaya) [inline crates/router/src/connector/nexinets/transformers.rs:630]: is this tested?

File: crates/router/src/connector/utils.rs [hunk 4]
```diff
@@ -399,7 +399,7 @@ impl WalletData for api::WalletData {
     fn get_wallet_token(&self) -> Result<String, Error> {
         match self {
             Self::GooglePay(data) => Ok(data.tokenization_data.token.clone()),
-            Self::ApplePay(data) => Ok(data.payment_data.clone()),
+            Self::ApplePay(data) => Ok(data.get_applepay_decoded_payment_data()?),
```
Comment (jagan-jaya) [inline crates/router/src/connector/utils.rs:402]: will this affect existing users?

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/nexinets/transformers.rs [hunk 5]
```diff
@@ -100,6 +138,29 @@ pub enum NexinetsBIC {
...
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum NexinetsWalletDetails {
+    ApplePayToken(Box<ApplePayDetails>),
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ApplePayDetails {
+    payment_data: Option<serde_json::Value>,
```
Comment (Sangamesh26) [inline crates/router/src/connector/nexinets/transformers.rs:150]: Why payment_data is option in applepay request?

Comment (srujanchikke) [reply crates/router/src/connector/nexinets/transformers.rs]: Yes , `payment_data` should not be an option , I have removed it and made it 'serde_json_Value' .

File: crates/router/src/connector/utils.rs (hunk 4 above)
Comment (srujanchikke) [reply crates/router/src/connector/utils.rs:402]: No ,`get_wallet_token_as_json` function for apple pay is only used in checkout and it has same use case as nexi nets .This is checked with @SamraatBansal  as well .

File: crates/router/src/connector/nexinets/transformers.rs (hunk 3 above)
Comment (srujanchikke) [reply crates/router/src/connector/nexinets/transformers.rs:621]: This is not tested . Although i have compared the request format after building request .

File: crates/router/src/connector/nexinets/transformers.rs [hunk 6]
```diff
@@ -525,13 +586,79 @@ fn get_payment_details_and_product(
...
-    NexinetsPaymentDetails::Card(Box::new(NexiCardDetails {
+fn get_card_data(
+    item: &types::PaymentsAuthorizeRouterData,
+    card: &api_models::payments::Card,
+) -> Result<NexinetsPaymentDetails, errors::ConnectorError> {
+    if item.request.is_mandate_payment() {
+        let cof_contract = Some(CofContract {
+            recurring_type: RecurringType::Unscheduled,
+        });
+        let card_data = if item.request.off_session.is_some() {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/nexinets/transformers.rs:597]: You should check for Some(true)

File: crates/router/src/connector/nexinets/transformers.rs [hunk 7]
```diff
@@ -525,13 +586,79 @@ fn get_payment_details_and_product(
     }
 }
 
-fn get_card_details(req_card: &api_models::payments::Card) -> NexinetsPaymentDetails {
-    NexinetsPaymentDetails::Card(Box::new(NexiCardDetails {
+fn get_card_data(
+    item: &types::PaymentsAuthorizeRouterData,
+    card: &api_models::payments::Card,
+) -> Result<NexinetsPaymentDetails, errors::ConnectorError> {
+    if item.request.is_mandate_payment() {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/nexinets/transformers.rs:593]: replace this with below code 
```suggestion
let (card_data, cof_contract) = match item.request.is_mandate_payment() {
        true => {
            let card_data = match item.request.off_session {
                Some(true) => CardDataDetails::PaymentInstrument(Box::new(PaymentInstrument {
                    payment_instrument_id: item.request.connector_mandate_id(),
                })),
                _ => CardDataDetails::CardDetails(Box::new(get_card_details(card))),
            };
            let cof_contract = Some(CofContract {
                recurring_type: RecurringType::Unscheduled,
            });
            (card_data, cof_contract)
        }
        false => (
            CardDataDetails::CardDetails(Box::new(get_card_details(card))),
            None,
        ),
    };
    Ok(NexinetsPaymentDetails::Card(Box::new(NexiCardDetails {
        card_data,
        cof_contract,
    })))
```

Lets not have early returns

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/utils.rs (hunk 4 above)
Comment (ArjunKarthik) [reply crates/router/src/connector/utils.rs:402]: Okay Cool.

**Review by ArjunKarthik — APPROVED** [review]

**Review by HyperSwitchers — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**